plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/launcher/WsAgentWithAuthLauncherImpl.java

```diff
@@ -0,0 +1,159 @@
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.UriBuilder;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+
+/**
+ * Starts the ws-agent and pings it using custom machine request,
+ * until ws-agent sends appropriate event about start.
+ *
+ * @author Anton Korneta
+ */
+@Singleton
+public class WsAgentWithAuthLauncherImpl implements WsAgentLauncher {
+    public static final String WS_AGENT_PROCESS_START_COMMAND = "machine.ws_agent.run_command";
+    public static final String WS_AGENT_PROCESS_NAME          = "CheWsAgent";
+
+    private static final Logger LOG                             = LoggerFactory.getLogger(WsAgentWithAuthLauncherImpl.class);
+    private static final String WS_AGENT_PROCESS_OUTPUT_CHANNEL = "workspace:%s:ext-server:output";
+
+    private final int                      wsAgentPingConnectionTimeoutMs;
+    private final long                     wsAgentMaxStartTimeMs;
+    private final long                     wsAgentPingDelayMs;
+    private final String                   wsAgentStartCommandLine;
+    private final String                   apiEndpoint;
+    private final String                   pingTimedOutErrorMessage;
+    private final Provider<MachineManager> machineManagerProvider;
+    private final HttpJsonRequestFactory   httpJsonRequestFactory;
+
+    @Inject
+    public WsAgentWithAuthLauncherImpl(Provider<MachineManager> machineManagerProvider,
+                                       HttpJsonRequestFactory httpJsonRequestFactory,
+                                       @Named(WS_AGENT_PROCESS_START_COMMAND) String wsAgentStartCommandLine,
+                                       @Named("machine.ws_agent.max_start_time_ms") long wsAgentMaxStartTimeMs,
+                                       @Named("machine.ws_agent.ping_delay_ms") long wsAgentPingDelayMs,
+                                       @Named("machine.ws_agent.ping_conn_timeout_ms") int wsAgentPingConnectionTimeoutMs,
+                                       @Named("machine.ws_agent.ping_timed_out_error_msg") String pingTimedOutErrorMessage,
+                                       @Named("api.endpoint") String apiEndpoint) {
+        this.wsAgentPingConnectionTimeoutMs = wsAgentPingConnectionTimeoutMs;
+        this.wsAgentMaxStartTimeMs = wsAgentMaxStartTimeMs;
+        this.wsAgentPingDelayMs = wsAgentPingDelayMs;
+        this.wsAgentStartCommandLine = wsAgentStartCommandLine;
+        this.pingTimedOutErrorMessage = pingTimedOutErrorMessage;
+        this.machineManagerProvider = machineManagerProvider;
+        this.httpJsonRequestFactory = httpJsonRequestFactory;
+        this.apiEndpoint = apiEndpoint;
+    }
+
+    public static String getWsAgentProcessOutputChannel(String workspaceId) {
+        return String.format(WS_AGENT_PROCESS_OUTPUT_CHANNEL, workspaceId);
+    }
+
+    @Override
+    public void startWsAgent(String workspaceId) throws NotFoundException, MachineException, InterruptedException {
```

Comment:
I think it is better to refactor original WsAgentLauncherImpl to simplify extension.